ddl/table.go

```diff
@@ -948,56 +956,188 @@ func updateVersionAndTableInfo(t *meta.Meta, job *model.Job, tblInfo *model.Tabl
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+
+		err = checkAddPartitionNameUnique(tblInfo, partInfo)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+		partDefPointers = updatePartitionInfo(partInfo, tblInfo)
+	}
+
+	originalState := partDefPointers[0].State
+	switch partDefPointers[0].State {
+	case model.StateNone:
+		// none -> delete only
+		job.SchemaState = model.StateDeleteOnly
+		for _, pdDef := range partDefPointers {
+			pdDef.State = model.StateDeleteOnly
+		}
+		ver, err = updateVersionAndTableInfoWithCheck(t, job, tblInfo, originalState != partDefPointers[0].State)
+	case model.StateDeleteOnly:
+		// delete only -> public
+		// Here need do some tiflash replica complement check.
+		if tblInfo.TiFlashReplica != nil && tblInfo.TiFlashReplica.Available {
+			// For available state, the new added partition should wait it's replica to
+			// be finished. Otherwise the query to this partition will be blocked.
+			needWait, err := checkPartitionReplica(partDefPointers, d, job)
+			if err != nil {
+				ver, err = convertAddTablePartitionJob2RollbackJob(t, job, err)
+				return ver, err
+			}
+			if needWait {
+				// The new added partition hasn't been replicated.
+				// Do nothing to the job this time, wait next worker round.
+				time.Sleep(tiflashCheckTiDBHttpAPIHalfInterval)
+				return ver, nil
+			}
+		}
+
+		// For normal table, change the state to public directly.
+		for _, pdDef := range partDefPointers {
+			pdDef.State = model.StatePublic
+		}
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, originalState != partDefPointers[0].State)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		// Finish this job.
+		job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+		asyncNotifyEvent(d, &util.Event{Tp: model.ActionAddTablePartition, TableInfo: tblInfo, PartInfo: partInfo})
+	default:
+		err = ErrInvalidDDLState.GenWithStackByArgs("partition", partDefPointers[0].State)
+	}
+
 	return ver, errors.Trace(err)
 }
 
-func updatePartitionInfo(partitionInfo *model.PartitionInfo, tblInfo *model.TableInfo) {
+func updatePartitionInfo(partitionInfo *model.PartitionInfo, tblInfo *model.TableInfo) []*model.PartitionDefinition {
 	parInfo := &model.PartitionInfo{}
 	oldDefs, newDefs := tblInfo.Partition.Definitions, partitionInfo.Definitions
 	parInfo.Definitions = make([]model.PartitionDefinition, 0, len(newDefs)+len(oldDefs))
 	parInfo.Definitions = append(parInfo.Definitions, oldDefs...)
 	parInfo.Definitions = append(parInfo.Definitions, newDefs...)
 	tblInfo.Partition.Definitions = parInfo.Definitions
+	partDefPointers := make([]*model.PartitionDefinition, 0, len(partitionInfo.Definitions))
+	for _, pdAdded := range partitionInfo.Definitions {
+		pdInMetaPointer := tblInfo.FindPartitionDefinitionByName(pdAdded.Name.L)
+		partDefPointers = append(partDefPointers, pdInMetaPointer)
+	}
+	return partDefPointers
```

Comment:
This part should be moved out of the function. It has nothing to do with updating partition information.

Reply:
It is convinent to return those new added partion pointer in the tableInfo.

Reply:
This reason doesn't make sense.

Reply:
I think it's ok here